Initial refactor of the documentation (#485)
qiboteam/qibocal: doc/source/developer-guides/new_protocol.rst

How to add a new protocol
=========================

In this tutorial we show how to add new protocol to ``Qibocal``.

Protocol implementation in ``Qibocal``
--------------------------------------

> Currently, characterization/calibration protocols in ``Qibocal`` are implemented
> by performing a clear separation between *input parameters*, *data acquired* and
> *results*.
> 
> These are then connected through the following functions:
> 
> * `acquisition` that receives as input `parameters` and outputs `data`
> * `fit` that receives as input `data` and outputs `results`
> * `plot` that receives as input `data` and `results` to visualize the protocol
> 

Reviewer: ```suggestion
Currently, characterization/calibration protocols are divided in three steps: `acquisition`, `fit` and `plot`. ``Qibocal`` provides three data structures  *input parameters*, *data acquired* and
*results*, that collect all the information concerning the routine.

The relationship between steps and data structures are summarized in the following bullets:

* `acquisition` receives as input `parameters` and outputs `data`
* `fit` receives as input `data` and outputs `results`
* `plot` receives as input `data` and `results` to visualize the protocol

```

> We believe that this approach is useful in order to code protocols that
> aims at achieving a specific task. However, this approach is flexible enough
> to show some data acquired without performing a post-processing analysis.

Reviewer: The first sentence is not clear, I suggest to drop it.
```suggestion
This approach is flexible enough to allow the data acquisition without performing a post-processing analysis.
```

All protocols are located in `src/qibocal/protocols/characterization <https://github.com/qiboteam/qibocal/tree/main/src/qibocal/protocols/characterization>`_.
Suppose that we want to code a protocol to perform a RX rotation for different
angles.

We create a file ``rotate.py`` in ``src/qibocal/protocols/characterization``.

Parameters
^^^^^^^^^^
First, we define the input parameters.

.. code-block:: python

    from dataclasses import dataclass
    from ...auto.operation import Parameters

    @dataclass
    class RotationParameters(Parameters):
        """Parameters for rotation protocol."""

        theta_start: float
        """Initial angle."""
        theta_end: float
        """Final angle."""
        theta_step: float
        """Angle step."""
        nshots: int
        """Number of shots."""

In this case you define a range for the angle to be probed alongside the number
of shots.

.. note::
      It is advised to use `dataclasses`. If you are not familiar
      have a look at the official `documentation <https://docs.python.org/3/library/dataclasses.html>`_.

Data structure
^^^^^^^^^^^^^^
Secondly, we define a data structure aims at storing both the angles and

> the probabilities measured for each qubit. A generic data structure is usually composed
> by some raw data (the data attribute), which is usually coded as a dictionary of arrays

Reviewer: ```suggestion
the probabilities measured for each qubit. A generic data structure is usually composed
of some raw data (the data attribute), which is usually coded as a dictionary of arrays
```

            platform (:class:`Platform`): Qibolab's platform
            qubits (dict): dict of target :class:`Qubit` objects to be characterized

        Returns:
            data (:class:`RotationData`)
        """

        # costruct range from RotationParameters
        angles = np.linspace(params.theta_start, params.theta_end, params.theta_step)
        # create data structure
        data = RotationData()

        # create and execute circuit for each angle
        for angle in angles:

            circuit = Circuit(platform.nqubits)
            for qubit in qubits:
                circuit.add(gates.RX(qubit, theta=angle))
                circuit.add(gates.M(qubit))

            result = circuit(nshots=params.nshots)

            for qubit in qubits:

                # extract probability of 0
                prob = result.probabilities(qubits=[qubit])[0]
                # store measurements in Rotation Data
                data.register_qubit(qubit, theta=angle, prob=prob)

        return data

Result class
^^^^^^^^^^^^

Here we decided to code a generic `Result` that contains the fitted
parameters for each quibt.

> .. code-block::python
> 
>     from qibolab.qubits import QubitId
> 
>     @dataclass
>     class RotationResults(Results):
>         """Results object for data"""
>         fitted_parameters: dict[QubitId, list] = field(default_factory=dict)
> 

Reviewer: ```suggestion
.. code-block:: python

    from qibolab.qubits import QubitId

    @dataclass
    class RotationResults(Results):
        """Results object for data"""
        fitted_parameters: dict[QubitId, list] = field(default_factory=dict)

```
